### PR TITLE
Speed up WildcardPattern.IsMatch if an pattern ends with asterisk

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Follow up on #10020 to add another fast path for `WildcardPattern.IsMatch` when an pattern ends with asterisk like `text*`.
In case of patterns like `text*`, `StartsWith(...)` method is used to check for if it's a match.

## PR Context

The main motivation was to complete the optimization of this code once we touched it.
However, this is a public code and should be improved.

As for the hot ways, my most important expectation was for path processing like `dir *`. In reality, test below shows a perf win of only about 1%.
```powershell
$a = dir -Recurse C:\Windows\System32\p*
```
Debug shows that globbing code does all works and then send filtered list child items to WindcardPattern.IsMatch() ! 
This way the engine almost always works with small collections and we don't see huge wins from fixes in WindcardPattern.IsMatch().

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/engine/regex.cs [hunk 1]
```diff
@@ -148,15 +148,8 @@ private void Init()
                 return;
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @iSazonov, your last commit had 527 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Validate that get-help works for AllUsers Scope.Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /Users/vsts/agent/2.153.2/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
Experimental Feature Basic Tests - Feature-Enabled.Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /Users/vsts/agent/2.153.2/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
help can be found for AllUsers Scope.Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /Users/vsts/agent/2.153.2/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
Validate Update-Help from the Web for all PowerShell modules..Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /Users/vsts/agent/2.153.2/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
Validate Update-Help from the Web for all PowerShell modules for user scope..Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /Users/vsts/agent/2.153.2/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @iSazonov, your last commit had 498 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Enable-ExperimentalFeature and Disable-ExperimentalFeature tests.Error occurred in Describe block
```powershell
Cannot find drive. A drive with the name 'TestRegistry' does not exist.
At D:\a\1\s\src\powershell-win-core\bin\Release\netcoreapp3.0\win7-x64\publish\Modules\Pester\4.8.0\Functions\TestRegistry.ps1:60 char:47
```
Validate about_help.txt under culture specific folder works.Error occurred in Describe block
```powershell
Cannot find drive. A drive with the name 'TestRegistry' does not exist.
At D:\a\1\s\src\powershell-win-core\bin\Release\netcoreapp3.0\win7-x64\publish\Modules\Pester\4.8.0\Functions\TestRegistry.ps1:60 char:47
```
Get-Help -Online opens the default web browser and navigates to the cmdlet help content.Error occurred in Describe block
```powershell
Cannot find drive. A drive with the name 'TestRegistry' does not exist.
At D:\a\1\s\src\powershell-win-core\bin\Release\netcoreapp3.0\win7-x64\publish\Modules\Pester\4.8.0\Functions\TestRegistry.ps1:60 char:47
```
Testing Get-Help Progress.Error occurred in Describe block
```powershell
Cannot find drive. A drive with the name 'TestRegistry' does not exist.
At D:\a\1\s\src\powershell-win-core\bin\Release\netcoreapp3.0\win7-x64\publish\Modules\Pester\4.8.0\Functions\TestRegistry.ps1:60 char:47
```
Language Primitive Tests.Error occurred in Describe block
```powershell
Cannot find drive. A drive with the name 'TestRegistry' does not exist.
At D:\a\1\s\src\powershell-win-core\bin\Release\netcoreapp3.0\win7-x64\publish\Modules\Pester\4.8.0\Functions\TestRegistry.ps1:60 char:47
```

**Comment by vexx32** [comment]
Would it also be worth mirroring this optimization for `.EndsWith()` i.e., `*pattern`?

File: src/System.Management.Automation/engine/regex.cs (hunk 1 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @iSazonov, your last commit had 527 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Clean up open Runspaces when exit powershell process.Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /home/vsts/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
Experimental Feature Basic Tests - Feature-Enabled.Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /home/vsts/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
Enable-ExperimentalFeature and Disable-ExperimentalFeature tests.Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /home/vsts/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
Validate Update-Help from the Web for all PowerShell modules..Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /home/vsts/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```
Language Primitive Tests.Error occurred in Describe block
```powershell
The type is not a known type for the file system. Only "file","directory" or "symboliclink" can be specified.
At /home/vsts/work/1/a/bins/publish/Modules/Pester/4.8.0/Functions/TestDrive.ps1:53 char:5
```

**Comment by iSazonov** [comment]
@vexx32 I found many places in our code where we use a pattern like "text*" so the PR will speed up the engine. But I don't see "*text" pattern. Although this is a public API and your proposal may be helpful for external applications.

**Comment by vexx32** [comment]
Yeah, I was more thinking just to speed up general usage. Definitely a good call to optimize the engine's usage, though! 💖

File: src/System.Management.Automation/engine/regex.cs [hunk 2]
```diff
@@ -171,8 +164,47 @@ private void Init()
...
+                    case '`':
+                        break;
+                    default:
+                        // No special characters present in the pattern, so we can just do a string comparison.
+                        _isMatch = str => string.Equals(str, Pattern, GetStringComparison());
+                        return;
+                }
+            }
+
+            if (Pattern.Length > 1 && Pattern.AsSpan().Slice(0, Pattern.Length - 1).IndexOfAny(s_specialChars) == -1)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Why not leverage the return value of `string.indexofany(char[])`?
> The zero-based index position of the first occurrence in this instance where any character in anyOf was found; -1 if no character in anyOf was found.

So we can do this:
```
if (Pattern.Length == 1 && Pattern[Pattern.Length - 1] == '*')
{
    ...
    return;
}

int index = Pattern.IndexOfAny(s_specialChars);
if (index == -1)
{
    ...
    return;
}
else if (index == Pattern.Length - 1 && Pattern[index] == '*')
{
    ...
    return;
}

var matcher = new WildcardPatternMatcher(this);
_isMatch = matcher.IsMatch;
```

File: src/System.Management.Automation/engine/regex.cs [hunk 3]
```diff
@@ -63,7 +63,7 @@ public sealed class WildcardPattern
         private static readonly Predicate<string> s_matchAll = _ => true;
 
         // wildcard pattern
-        internal string Pattern { get; }
+        internal string Pattern { get; private set; }
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: I don't see the setter being used anywhere.

File: src/System.Management.Automation/engine/regex.cs [hunk 4]
```diff
@@ -148,15 +148,8 @@ private void Init()
                 return;
             }
 
-            if (Pattern.Length == 1 && Pattern[0] == '*')
+            StringComparison GetStringComparison()
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Please move this local function to the beginning of this method.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
I would like to see some numbers that showcase the improvement of this change.
I think for all perf PRs, we should have some sort of evidence to show the effectiveness of the changes (make sure we are guided by profiling).
In case you don't have a specific scenario that gets noticeable improvement, at least show what code paths will hit this change and why it's a change that we want to have.

File: src/System.Management.Automation/engine/regex.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

**Comment by iSazonov** [comment]
The main motivation was to complete the optimization of this code once we touched it.
However, this is a public code and should be improved.

As for the hot ways, my most important expectation was for path processing like `dir *`. In reality, test below shows a perf win of only about 1%.
```powershell
$a = dir -Recurse C:\Windows\System32\p*
```
Debug shows that globbing code does all works and then send filtered list child items to WindcardPattern.IsMatch() ! 
This way the engine almost always works with small collections and we don't see huge wins from fixes in WindcardPattern.IsMatch().

File: src/System.Management.Automation/engine/regex.cs [hunk 5]
```diff
@@ -171,7 +159,33 @@ private void Init()
...
+            {
+                // No special characters present in the pattern, so we can just do a string comparison.
+                _isMatch = str => string.Equals(str, Pattern, GetStringComparison());
+                return;
+            }
+
+            if (index == Pattern.Length - 1 && Pattern[index] == '*')
+            {
+                // No special characters present in the pattern before last position and last character is asterisk.
+                var patternWithoutAsterisk = Pattern.Substring(0, Pattern.Length - 1);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Minor: replace `Pattern.Length - 1` with `index`.

**Comment by daxian-dbw** [comment]
Can you please move your comment in https://github.com/PowerShell/PowerShell/pull/10054#issuecomment-508384193 to the PR description?
Also, can you elaborate more about your following statements? I'm not very clear what you mean by, for example, `works with small collections`.
> Debug shows that globbing code does all works and then send filtered list child items to WindcardPattern.IsMatch() !
This way the engine almost always works with small collections and we don't see huge wins from fixes in WindcardPattern.IsMatch().

File: src/System.Management.Automation/engine/regex.cs (hunk 5 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Minor: replace `Pattern.Length - 1` with `index`.

**Review by daxian-dbw — APPROVED** [review]

File: src/System.Management.Automation/engine/regex.cs (hunk 5 above)
Comment (lzybkr) [reply src/System.Management.Automation/engine/regex.cs]: Can you avoid the string copy?

Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: We could do:
```c#
                _isMatch = str => str.AsSpan().StartsWith(Pattern.AsSpan().Slice(0, index), GetStringComparison());
```
In the code it seems we do more work - convert to span both string and slice Pattern string in every call. Also we can not pass the Span (ref local) to the anonymous method.
```c#
                var patternWithoutAsterisk = Pattern.AsMemory().Slice(0, index);
                _isMatch = str => str.AsSpan().StartsWith(patternWithoutAsterisk.Span, GetStringComparison());
```
I don't know what is better.

Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: I measured both cases and second proposal is nice:

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.557 (1809/October2018Update/Redstone5)
Intel Core i3-4150 CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT


```
|       Method | pattern |          text |     Mean |    Error |   StdDev |   Median | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |-------- |-------------- |---------:|---------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
|     **TestOrig** |      **a*** | **abcdefghijklm** | **169.7 ns** | **3.417 ns** | **7.501 ns** | **166.6 ns** |  **1.00** |    **0.00** | **0.1018** |     **-** |     **-** |     **160 B** |
| TestAsMemory |      a* | abcdefghijklm | 163.1 ns | 3.765 ns | 4.336 ns | 161.8 ns |  0.94 |    0.04 | 0.0918 |     - |     - |     144 B |
|              |         |               |          |          |          |          |       |         |        |       |       |           |
|     **TestOrig** |    **abc*** | **abcdefghijklm** | **196.6 ns** | **3.546 ns** | **2.961 ns** | **195.8 ns** |  **1.00** |    **0.00** | **0.1070** |     **-** |     **-** |     **168 B** |
| TestAsMemory |    abc* | abcdefghijklm | 194.1 ns | 4.193 ns | 7.453 ns | 191.2 ns |  1.00 |    0.04 | 0.0918 |     - |     - |     144 B |
|              |         |               |          |          |          |          |       |         |        |       |       |           |
|     **TestOrig** |  **abcde*** | **abcdefghijklm** | **226.8 ns** | **2.748 ns** | **2.294 ns** | **226.6 ns** |  **1.00** |    **0.00** | **0.1070** |     **-** |     **-** |     **168 B** |
| TestAsMemory |  abcde* | abcdefghijklm | 224.0 ns | 5.349 ns | 7.141 ns | 221.1 ns |  0.99 |    0.04 | 0.0918 |     - |     - |     144 B |

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/regex.cs]: Thanks @iSazonov for measuring the diffs

**Review by daxian-dbw — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@iSazonov The PR looks good, but can you please respond to my comment here: https://github.com/PowerShell/PowerShell/pull/10054#issuecomment-508808988
I would like to understand why this change only show 1% improvement to `dir -Recurse C:\Windows\System32\p*`.

**Comment by iSazonov** [comment]
@daxian-dbw When we send a path like `C:\Windows\System32\p*` to engine our globbing code does all work - parse and expand `p*` to a collection with leaves. Then the already filtered leaves (in the example, all started with "p") is passed to a code which again filters the leaves by WindcardPattern.IsMatch(). So (1) we don't filter large, all full path list, only small collection of leaves  - that's why we get so little acceleration, (2) we do duplicated work that we would address, I believe it is common problem in globbing code - there is many places where it happens.

**Comment by daxian-dbw** [comment]
@iSazonov I opened #10093 to track this, so we don't lose your analysis.

**Pull request merged**